contracts/automation/andromeda-task-balancer/src/contract.rs

```diff
@@ -0,0 +1,428 @@
+use crate::state::{State, STATE, STORAGE_CONTRACTS, UP_NEXT};
+use ado_base::state::ADOContract;
+use andromeda_automation::storage::ExecuteMsg as StorageExecuteMsg;
+use andromeda_automation::storage::InstantiateMsg as StorageInstantiateMsg;
+use andromeda_automation::storage::QueryMsg as StorageQueryMsg;
+use andromeda_automation::task_balancer::{
+    ExecuteMsg, GetSizeResponse, GetStorageResponse, InstantiateMsg, MigrateMsg, QueryMsg,
+};
+use common::{ado_base::InstantiateMsg as BaseInstantiateMsg, encode_binary, error::ContractError};
+use cosmwasm_std::{
+    ensure, entry_point, to_binary, Binary, CosmosMsg, Deps, DepsMut, Env, MessageInfo,
+    QueryRequest, Reply, Response, StdError, SubMsg, Uint128, WasmMsg, WasmQuery,
+};
+use cw2::{get_contract_version, set_contract_version};
+use semver::Version;
+use std::env;
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:andromeda-task-balancer";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    let state = State {
+        contracts: Uint128::zero(),
+        max: msg.max,
+        storage_code_id: msg.storage_code_id,
+        admin: info.sender.to_string(),
+    };
+
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    STATE.save(deps.storage, &state)?;
+    ADOContract::default().instantiate(
+        deps.storage,
+        env,
+        deps.api,
+        info,
+        BaseInstantiateMsg {
+            ado_type: "task-balancer".to_string(),
+            ado_version: CONTRACT_VERSION.to_string(),
+            operators: None,
+            modules: None,
+            primitive_contract: None,
+        },
+    )
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn reply(deps: DepsMut, _env: Env, msg: Reply) -> Result<Response, ContractError> {
+    let response = msg.result;
+    // We only get a reply on success, so it's safe to assume there's an event
+    // There's also only one event resulting from instantiation, so we access the first (and only) event
+    let address = &response.unwrap().events[0];
+
+    // According to the raw logs, the first key value pair holds the instantiated contract's address
+    let attribute = &address.attributes[0];
```

Comment:
There is a method in our package to retrieve the address from a reply if you'd prefer to use that